### PR TITLE
Fix option names

### PR DESCRIPTION
**Pull request by alinabuzachis**
##### SUMMARY

Add the original option name as an alias to avoid issues when snake cased option names are not correctly reversed to the original CamelCase. Use the alias via the CloudControl API rather than performinng that CamelCase transformations.

Collection re-generated using https://github.com/ansible-community/ansible.content_builder/pull/62

##### ISSUE TYPE

- Bugfix Pull Request

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: plugins/module_utils/core.py [hunk 1]
```diff
@@ -255,22 +265,24 @@ def present(
...
             resource = self.client.get_resource(
                 TypeName=type_name, Identifier=identifier, aws_retry=True
             )
             results = self.update_resource(resource, params, create_only_params)
         except self.client.exceptions.ResourceNotFoundException:
             if self.module.params.get("identifier"):
+                _primary_identifier = [id.split("/")[-1] for id in primary_identifier]
                 self.module.fail_json(
-                    f"""You must specify both {*primary_identifier, } to create a new resource.
+                    f"""You must specify together {*_primary_identifier, } to create a new resource.
```
Comment (gravesm) [inline plugins/module_utils/core.py:285]: I'm not sure I understand this change. Is it because there may be more than two?

File: plugins/module_utils/utils.py [hunk 2]
```diff
@@ -106,7 +106,11 @@ def scrub_keys(a_dict: Dict, list_of_keys_to_remove: List[str]) -> Dict:
     """Filter a_dict by removing unwanted key: values listed in list_of_keys_to_remove"""
     if not isinstance(a_dict, dict):
         return a_dict
-    return {k: v for k, v in a_dict.items() if k not in list_of_keys_to_remove}
+    return {
+        k: v
+        for k, v in a_dict.items()
+        if not any(k in item for item in list_of_keys_to_remove)
```
Comment (gravesm) [inline plugins/module_utils/utils.py:112]: What was the reason for this change?

File: tests/sanity/ignore-2.10.txt
Comment (gravesm) [inline tests/sanity/ignore-2.10.txt:1]: We should be able to delete the ignore files for 2.10, 2.11 and 2.12.

Comment (alinabuzachis) [reply tests/sanity/ignore-2.10.txt:1]: Done!

File: plugins/module_utils/utils.py (hunk 2 above)
Comment (alinabuzachis) [reply plugins/module_utils/utils.py]: Reverted!

File: plugins/module_utils/core.py (hunk 1 above)
Comment (alinabuzachis) [reply plugins/module_utils/core.py:284]: There might be, that's why I wanted to make it more generic.

File: plugins/module_utils/utils.py [hunk 3]
```diff
@@ -264,6 +299,24 @@ def op(operation, path, value):
...
+
+    for dict2 in list2:
+        matching_dicts = [
+            dict1
+            for dict1 in merged_list
+            if all(item in dict1.items() for item in dict2.items())
+        ]
+        if matching_dicts:
+            for k, v in dict2.items():
+                matching_dicts[0][k] = v
```
[the remark this answers is not in the document]
Comment (alinabuzachis) [reply plugins/module_utils/utils.py]: It was an outdated code for merge_dicts. Thanks! Updated the code and added a unit test.

File: plugins/module_utils/core.py (hunk 1 above)
Comment (gravesm) [reply plugins/module_utils/core.py:284]: OK, I would suggest a change to wording in that case. Something like `You must specify all of the following to create a new resource: {*_primary_identifier, }.`

Comment (alinabuzachis) [reply plugins/module_utils/core.py]: Done, thanks!

File: plugins/module_utils/utils.py [hunk 4]
```diff
@@ -264,6 +297,49 @@ def op(operation, path, value):
...
+
+    for dict2 in list2:
+        matching_dicts = [
+            dict1
+            for dict1 in merged_list
+            if dict1.get("PolicyName") == dict2.get("PolicyName")
+        ]
+
+        if matching_dicts:
+            dict1 = matching_dicts[0]
```
Comment (gravesm) [inline plugins/module_utils/utils.py:311]: What happens if there is more than one matching dict?

File: tests/integration/targets/memorydb/aliases [hunk 5]
```diff
@@ -1 +1,4 @@
 cloud/aws
+
+# Missing permissions
+disabled
```
Comment (gravesm) [inline tests/integration/targets/memorydb/aliases:4]: Will https://github.com/mattclay/aws-terminator/pull/291 not allow us to run these tests?

Comment (alinabuzachis) [reply tests/integration/targets/memorydb/aliases:4]: I disabled the tests because at that time there wasn't a PR with the terminator policies. I assume (but have to test), tests will pass.

File: plugins/module_utils/utils.py [hunk 6]
```diff
@@ -264,6 +297,49 @@ def op(operation, path, value):
...
+            for key, value in dict2.items():
+                try:
+                    value_dict = ast.literal_eval(value)
```
[the remark this answers is not in the document]
Comment (gravesm) [reply plugins/module_utils/utils.py:310]: Isn't the policy document JSON, not a python dict? `json.loads()` would be a better option if this is the case.

[the remark this answers is not in the document]
Comment (alinabuzachis) [reply plugins/module_utils/utils.py]: Yeah, done!

File: plugins/module_utils/utils.py [hunk 7]
```diff
@@ -264,6 +297,51 @@ def op(operation, path, value):
...
 
+def safe_json(data: str) -> Dict:
+    try:
+        json_object = json.loads(data)
+    except (ValueError, TypeError) as e:
+        return data
+    return json_object
+
+
+def merge_dicts(list1: List, list2) -> List:
```
Comment (gravesm) [inline plugins/module_utils/utils.py:308]: What's the problem that this function is trying to solve? Is there an example of a resource property that needs this? It makes some assumptions about the structure of the arguments that at the very least should probably be documented. I'm worried that those assumptions may not always be true. For example, the following will raise an exception:

```python
old = [
    {"foo": "bar"}
]
new = [
    {"foo": "gaz", "someotherkey": {"blah": "asdf"}}
]
merge_dicts(old, new)
```

Comment (alinabuzachis) [reply plugins/module_utils/utils.py:309]: @gravesm I fixed that corner case and add a test case for it. However, this function handles the case when two list of dictionaries (with keys different than Key) needs to merged. For example:

- I create in a first instance the role below and then try to re-run the same task (because I want to check idempotency of update the role) it will fail because it will be unable to compare list of dicts inside `policies` parameter.
```
    - name: Create minimal lambda role
      amazon.cloud.iam_role:
        role_name: '{{ lambda_role_name }}'
        assume_role_policy_document: '{{ lookup("file", "lambda-minimal-policy.json") }}'
        policies:
          - policy_name: "{{ lambda_policy_name }}"
            policy_document: '{{ lookup("file", "lambda-policy.json") }}'
        managed_policy_arns:
          - "arn:aws:iam::{{ _caller_info.account }}:policy/{{ lambda_policy_name }}"
        wait: true
      register: _create_iam_role_result 
```

Also, policy_document returned by the API will be a dictionary while the one the user will specify will be a string, so, that's why I created safe_load().

We should also ensure the right quotes are used within the policy_document and that's why I had to add `QuoteSwappingEncoder`. The test case I added covers exactly the case where we need this quotes swapping.

File: plugins/module_utils/core.py [hunk 8]
```diff
@@ -286,7 +301,8 @@ def present(
 
     def create_resource(self, type_name: str, params: Dict) -> bool:
         changed: bool = False
-        params = json.dumps(params)
+
+        params = json.dumps(params, cls=QuoteSwappingEncoder)
```
Comment (gravesm) [inline plugins/module_utils/core.py:305]: This creates another problem. Consider the following task:

```yaml
    - amazon.cloud.iam_role:
        assume_role_policy_document:
          Version: 2012-10-17
          Statement:
            - Effect: Deny
              Principal:
                AWS: "*"
              Action: sts:AssumeRole
        description: "mgraves's awesome role"
        role_name: mgraves-awesome-role
```
Any string field that allows you to use single or double quotes is going to swap them, whether it makes any sense or not. I feel like we are in a vicious cycle where each change you make causes a new bug. I would suggest taking a step back and thinking about what problem it is you are trying to solve. It sounds like the point of the quote swapping and the  `merge_dicts()` logic is to merge IAM policy documents. If that's the case, then create a set of changes that is focused on doing only that. You won't have to deal with so many corner cases. The problem now is that you are trying to create something that can work on any dictionary, which is going to be very difficult to do correctly. It may even be impossible since there may be different behaviors needed for different resource properties.

Comment (alinabuzachis) [reply plugins/module_utils/core.py:305]: `merge_dicts()` is needed for the reason I explained before. It's not only tied to iam_role. The question of swapping quotes, I can isolate it to only a Policy parameter with a PolicyDocument suboption.

Comment (gravesm) [reply plugins/module_utils/core.py:305]: > merge_dicts() is needed for the reason I explained before. It's not only tied to iam_role.

But the reason you gave was specifically tied to policies. Are there other cases?

**Comment by alinabuzachis** [comment]
The same issue, the CI does not use the fork's branch, but the main branch, that's why the eke tests are still failing.

File: plugins/module_utils/utils.py [hunk 9]
```diff
@@ -264,6 +299,205 @@ def op(operation, path, value):
...
+                if isinstance(merged_list, list):
+                    dict1 = merged_list[index]
+                    if isinstance(dict2, dict) and isinstance(dict1, dict):
+                        for key, value in dict2.items():
+                            value_dict = safe_json(value)
+                            if (
+                                dict1.get(key, {})
+                                and isinstance(value_dict, dict)
+                                and isinstance(dict1[key], dict)
+                            ):
```
Comment (gravesm) [inline plugins/module_utils/utils.py:407]:
```suggestion
                            if (
                                isinstance(value_dict, dict)
                                and isinstance(dict1.get(key), dict)
                            ):
```

File: plugins/module_utils/utils.py [hunk 10]
```diff
@@ -264,6 +299,205 @@ def op(operation, path, value):
...
+                                dict1[key] = recursive_merge(dict1[key], value_dict)
+                            elif (
+                                dict1.get(key, [])
+                                and isinstance(value_dict, list)
+                                and isinstance(dict1[key], list)
+                            ):
```
Comment (gravesm) [inline plugins/module_utils/utils.py:413]:
```suggestion
                            elif (
                                isinstance(value_dict, list)
                                and isinstance(dict1.get(key), list)
                            ):
```

File: plugins/module_utils/utils.py [hunk 11]
```diff
@@ -264,6 +299,205 @@ def op(operation, path, value):
...
+                                # Easy merge
+                                if all(
+                                    isinstance(item, (str, int, float))
```
Comment (gravesm) [inline plugins/module_utils/utils.py:416]: Do we care about `bool`? A better type comparison here would probably be to use `collections.abc.Hashable` because that's what we really care about.
